File: bridge/python/reflect.py
"""What a dspy signature looks like to the crate: one plain description per field.

Reading a type off a Python annotation is reflection, and only Python can do it. Deciding how
the result reads in a prompt is rendering, and only the crate does that. This module is the
line between: every function here answers a question about an annotation and answers it in
data, so nothing on this side ever formats a byte the model will see.

A question this side cannot answer is raised rather than guessed at. `Unsupported` derives from
`BaseException` so it lands outside every `except Exception` dspy runs, because a case rendered
on Python would report a pass for code this crate has not written.
"""

from __future__ import annotations

import enum
import json
import types
import typing

import pydantic
from dspy.adapters.types.base_type import Type
from dspy.adapters.types.code import Code
from dspy.adapters.utils import (
    _annotation_is_subclass,
    _get_json_schema,
    get_annotation_name,
)

# The Rust FieldKind each Python annotation maps to. Anything absent is not yet modelled.
KINDS: dict[typing.Any, str] = {str: "str", int: "int", float: "float", bool: "bool"}


class Unsupported(BaseException):
    """The case needs a Rust feature that does not exist yet.

    Deriving from `BaseException` puts this outside `except Exception`, which is what dspy's
    JSON-adapter fallback catches on both paths. Inheriting from `Exception` would leave the
    exclusion resting on a matching `except Unsupported: raise` in every handler that could
    ever see one, and the async path already inherits a handler this shim does not write.
    """


def kind_of(annotation: typing.Any) -> str:
    """A scalar names itself; anything else carries the name dspy would print.

    Sending `json:<annotation>` rather than a bare `json` is what lets the numbered line read
    `(PetOwner)` the way dspy renders it, instead of collapsing every non-scalar to one word.
    """
    if typing.get_origin(annotation) is typing.Literal:
        # The closed set travelling beside this becomes the printed annotation, so what is
        # named here is the wire type under it: text, which is how every member of a Literal
        # reaches the marker path whatever its Python type.
        return "str"
    try:
        return KINDS[annotation]
    except (KeyError, TypeError):
        pass
    # Ahead of the JSON kinds: pydantic can schema an enum, but dspy does not describe one that
    # way — it names the type and lists its members' values.
    if isinstance(annotation, type) and issubclass(annotation, enum.Enum):
        return f"enum:{get_annotation_name(annotation)}"
    if _carries_as_json(annotation):
        return f"json:{get_annotation_name(annotation)}"
    raise Unsupported(f"no Rust FieldKind for annotation {annotation!r}")


def _carries_as_json(annotation: typing.Any) -> bool:
    """Whether the crate's `Json` kind can carry values of this annotation.

    A model does, since its values are objects. A container does exactly when everything it
    holds does — `list[Tool]` rides on `Tool` — because the container is JSON either way and
    what is inside it is what has to survive the crossing. Anything else says so rather than
    crossing as an annotation whose values were never checked: a kind that renders a value
    wrongly is worse than one that refuses it, because the prompt still looks plausible.
    """
    if isinstance(annotation, type) and issubclass(annotation, pydantic.BaseModel):
        return True
    # A plain class pydantic can describe carries too — `datetime` reaches the model as the
    # string its schema says it is. Asking pydantic is the same question the crate's schema
    # note asks later, so a type that answers it renders consistently.
    if isinstance(annotation, type) and not typing.get_args(annotation):
        try:
            pydantic.TypeAdapter(annotation).json_schema()
            return True
        except Exception:
            return False
    args = typing.get_args(annotation)
    return bool(args) and all(
        arg is Ellipsis or arg is type(None) or _scalar_or_json(arg) for arg in args
    )


def _scalar_or_json(annotation: typing.Any) -> bool:
    """Whether one member of a container is itself carryable."""
    try:
        return annotation in KINDS or _carries_as_json(annotation)
    except TypeError:
        return _carries_as_json(annotation)


def closed_set_of(annotation: typing.Any) -> str | None:
    """A `Literal`'s members as JSON, or None where the annotation is not one.

    JSON is what the bridge carries a closed set over, and it spells the three member types
    Rust models. A `Literal` over anything else — an Enum, None, bytes — has no crossing yet,
    and must say so rather than lose members on the way across.
    """
    # An enum's members are its closed set, carried as the values dspy asks the model for.
    if isinstance(annotation, type) and issubclass(annotation, enum.Enum):
        return json.dumps([member.value for member in annotation])
    if typing.get_origin(annotation) is not typing.Literal:
        return None
    members = typing.get_args(annotation)
    # A member Python prints as itself — an enum member — crosses as its `str`, tagged so the
    # crate spells it bare rather than quoting it into something the model would answer wrong.
    return json.dumps(
        [
            member if isinstance(member, (str, int, bool)) else {"bare": str(member)}
            for member in members
        ]
    )


def schema_of(kind: str, annotation: typing.Any) -> str | None:
    """A structured field's JSON schema, as dspy builds it, or None for a scalar.

    Reading a schema off a Python annotation is pydantic's job, so upstream's own extractor runs
    here — key order included, since it is part of the bytes. Whether the schema reaches the
    prompt stays the crate's decision: a type whose description already states its contract
    drops the note, and the crate is what knows that.

    Only a structured field is asked for one, which is where the crate consults it and where
    dspy computes it. An annotation that cannot produce one is a gap in this bridge rather than
    a field to render blank, so it says so instead of quietly dropping the schema — a missing
    note renders a prompt that looks right and is not.
    """
    if not kind.startswith("json:"):
        return None
    try:
        return json.dumps(_get_json_schema(annotation), ensure_ascii=False)
    except Exception as error:
        raise Unsupported(f"no JSON schema for annotation {annotation!r}: {error}") from error


def type_descriptions_of(annotation: typing.Any) -> str | None:
    """The custom types an annotation names, as JSON `[[name, prose], ...]`, or None.

    Which types an annotation mentions is Python reflection, so it happens here; how the pairs
    read in a prompt is the crate's business, so only the pairs cross.
    """
    described = [
        {
            "name": get_annotation_name(custom),
            "text": custom.description(),
            # dspy asks whether the annotation *is* a `dspy.Code`, not what it is called, so
            # this asks the same way: a subclass counts and a look-alike does not.
            "replaces_schema": _annotation_is_subclass(custom, Code),
        }
        for custom in Type.extract_custom_type_from_annotation(annotation)
        if custom.description()
    ]
    return json.dumps(described) if described else None


def reflection_of(kind: str, annotation: typing.Any) -> str | None:
    """A structured field's own shape, as JSON, or None for a scalar.

    An adapter that states the declared type itself rather than a schema of it needs facts a
    schema throws away: a member keyed by its name *and* the alias it answers to elsewhere,
    the difference between `object` and a type with no constraints, a mapping's key type, and
    every model's docstring. Those come off pydantic, so they are read here, as the tree dspy's
    own renderer walks — with nothing walked into text, which is the crate's half.

    Models cross as a table referred to by index, so a type that names itself arrives whole
    instead of recursing forever on this side. Whether such a type can be rendered at all is
    then the crate's call, made where dspy makes it.
    """
    if not kind.startswith("json:"):
        return None
    models: list = []
    seen: dict[type, int] = {}
    return json.dumps({"type": _node(annotation, models, seen), "models": models})


def _node(annotation: typing.Any, models: list, seen: dict) -> dict:
    """One node of an annotation, in the same order of questions dspy's renderer asks them."""
    for scalar, kind in KINDS.items():
        if annotation is scalar:
            return {"kind": kind}
    if _annotation_is_subclass(annotation, pydantic.BaseModel):
        return {"kind": "model", "model": _model(annotation, models, seen)}
    try:
        origin, args = typing.get_origin(annotation), typing.get_args(annotation)
    except Exception:
        return _named(annotation)
    if origin in (types.UnionType, typing.Union):
        present = [arg for arg in args if arg is not type(None)]
        return {
            "kind": "union",
            "of": [_node(arg, models, seen) for arg in present],
            "optional": len(present) < len(args),
        }
    if origin is typing.Literal:
        return _literal(annotation, args)
    if origin is list:
        return {"kind": "list", "of": _node(args[0], models, seen)}
    if origin is dict:
        return {
            "kind": "dict",
            "key": _node(args[0], models, seen),
            "value": _node(args[1], models, seen),
        }
    return _named(annotation)


def _model(model: type, models: list, seen: dict) -> int:
    """This model's slot in the table, filling it the first time the model is reached.

    The slot is claimed before the members are walked, so a model that names itself refers back
    to the entry being built rather than recursing here forever.
    """
    if model in seen:
        return seen[model]
    index = seen[model] = len(models)
    models.append(None)
    models[index] = {
        "doc": model.__doc__,
        "fields": [
            {
                "name": name,
                "desc": field.description,
                "alias": field.alias,
                "type": _node(field.annotation, models, seen),
            }
            for name, field in model.model_fields.items()
        ],
    }
    return index


def _literal(annotation: typing.Any, members: tuple) -> dict:
    """A closed set as its members, which the crate spells the way Python's own `str` does.

    A member JSON has no spelling for — an Enum, bytes — would have to cross as text this side
    had already formatted, so the whole type falls back to its printed name instead.
    """
    if all(member is None or isinstance(member, (str, int, bool)) for member in members):
        return {"kind": "literal", "members": list(members)}
    return _named(annotation)


def _named(annotation: typing.Any) -> dict:
    """dspy's last resort, and this one: the name Python prints for the type."""
    name = annotation.__name__ if hasattr(annotation, "__name__") else str(annotation)
    return {"kind": "named", "name": name}


def describe(fields: dict) -> list[tuple]:
    described = []
    for name, info in fields.items():
        desc = info.json_schema_extra.get("desc") or ""
        if desc == f"${{{name}}}":  # dspy's placeholder for "no description given"
            desc = ""
        annotation = info.annotation
        kind = kind_of(annotation)
        described.append(
            (
                name,
                kind,
                desc,
                closed_set_of(annotation),
                type_descriptions_of(annotation),
                reflection_of(kind, annotation),
            )
        )
    return described


def described_outputs(signature) -> list[tuple]:
    """Outputs carry the nested schema of a structured field ahead of the closed set."""
    return [
        (
            name,
            kind,
            desc,
            schema_of(kind, signature.output_fields[name].annotation),
            values,
            types_named,
            reflection,
        )
        for name, kind, desc, values, types_named, reflection in describe(signature.output_fields)
    ]

File: crates/dsrust/src/signature/reflect.rs
//! A declared type's own structure, read off its JSON schema.
//!
//! [`BamlAdapter`](crate::BamlAdapter) states a type rather than a schema of it — `title: string`
//! over four lines instead of a `{"type": "object", "properties": …}` blob — and to do that it
//! needs the shape, not the schema. dspy gets it by walking the pydantic annotation at run time.
//! Rust has no run-time annotation to walk, so it is taken from the one description of the type
//! that already exists: the `schemars` schema every signature field is required to have.
//!
//! Without this a Rust-declared signature reached BAML as the bare word `json`, which is the one
//! thing that adapter exists not to send.
//!
//! The tree this builds is the same one `crates/dsrs-bridge/python/reflect.py` produces from a pydantic model,
//! because [`baml::notation`](crate::adapter::baml) reads exactly one shape whichever side built
//! it.

use serde_json::{Value, json};

use super::json_field_schema;

/// The structure of `T`, in the form an adapter that states types can render.
pub fn json_field_reflection<T: schemars::JsonSchema>() -> Value {
    from_schema(&json_field_schema::<T>())
}

/// The same, for a schema already in hand.
fn from_schema(schema: &Value) -> Value {
    let mut models = Vec::new();
    let declared = node(schema, &mut models);
    json!({ "type": declared, "models": models })
}

/// One node of the tree, pushing any object it meets onto `models` and referring to it by index.
fn node(schema: &Value, models: &mut Vec<Value>) -> Value {
    if let Some(members) = schema.get("enum").and_then(Value::as_array) {
        return json!({ "kind": "literal", "members": members });
    }
    if let Some(arms) = union_arms(schema) {
        return union(arms, models);
    }
    match declared_type(schema) {
        Some("string") => json!({ "kind": "str" }),
        Some("integer") => json!({ "kind": "int" }),
        Some("number") => json!({ "kind": "float" }),
        Some("boolean") => json!({ "kind": "bool" }),
        Some("array") => json!({ "kind": "list", "of": items(schema, models) }),
        Some("object") => object(schema, models),
        // A schema that names no type describes anything — `serde_json::Value`, and what
        // pydantic spells `Any`. Upstream prints the name rather than pretending to a shape.
        _ => json!({ "kind": "named", "name": "any" }),
    }
}

/// A schema's `type`, ignoring the null arm of an optional — that is read by [`union_arms`].
fn declared_type(schema: &Value) -> Option<&str> {
    match &schema["type"] {
        Value::String(name) => Some(name.as_str()),
        Value::Array(names) => names
            .iter()
            .filter_map(Value::as_str)
            .find(|name| *name != "null"),
        _ => None,
    }
}

/// The arms of a choice, however the schema spells one: `anyOf`, `oneOf`, or a `type` array.
///
/// `None` when the schema is not a choice at all, which is what keeps a plain `{"type": "string"}`
/// from being wrapped in a one-armed union nobody wrote.
fn union_arms(schema: &Value) -> Option<Vec<Value>> {
    for key in ["anyOf", "oneOf"] {
        if let Some(arms) = schema[key].as_array() {
            return Some(arms.clone());
        }
    }
    let names = schema["type"].as_array()?;
    let arms: Vec<Value> = names
        .iter()
        .filter_map(Value::as_str)
        .map(|name| json!({ "type": name }))
        .collect();
    (arms.len() > 1).then_some(arms)
}

/// A choice, with the null arm lifted out — upstream renders `Optional[T]` as `T or null`
/// rather than as a union with a null in it.
fn union(arms: Vec<Value>, models: &mut Vec<Value>) -> Value {
    let optional = arms.iter().any(is_null);
    let of: Vec<Value> = arms
        .iter()
        .filter(|arm| !is_null(arm))
        .map(|arm| node(arm, models))
        .collect();
    // `Option<T>` needs no case of its own: a two-armed choice with a null arm leaves one type in
    // `of` and `optional` true, which is what the general form already writes. It *had* one, and
    // the two arms were byte-identical — which is why both mutants of its guard survived, and how
    // a branch that can never be told from its neighbour announces itself.
    json!({ "kind": "union", "of": of, "optional": optional })
}

fn is_null(schema: &Value) -> bool {
    schema["type"] == json!("null")
}

fn items(schema: &Value, models: &mut Vec<Value>) -> Value {
    match schema.get("items") {
        Some(items) => node(items, models),
        // An array that does not say what it holds holds anything.
        None => json!({ "kind": "named", "name": "any" }),
    }
}

/// An object is either a record with declared members or a map keyed by string, and the two are
/// told apart the way serde writes them: a map has values under `additionalProperties` and no
/// members of its own.
fn object(schema: &Value, models: &mut Vec<Value>) -> Value {
    let properties = schema.get("properties").and_then(Value::as_object);
    if properties.is_none() {
        // No `is_object` guard: `node` answers "any" for a schema that declares no type, which a
        // bare `true`/`false` `additionalProperties` is — so the guard could not change an answer
        // and its mutant survived. Letting `node` decide removes the branch rather than testing it.
        let value = schema.get("additionalProperties").map_or_else(
            || json!({ "kind": "named", "name": "any" }),
            |values| node(values, models),
        );
        // serde keys every map it writes by a string, so there is no other key type to find.
        return json!({ "kind": "dict", "key": { "kind": "str" }, "value": value });
    }
    let at = model(schema, models);
    json!({ "kind": "model", "model": at })
}

/// Record a model and answer with its index.
///
/// The slot is taken before the members are walked, so a type reaching itself refers back to the
/// index it already has rather than recursing forever. The renderer stops such a cycle where it
/// finds it.
fn model(schema: &Value, models: &mut Vec<Value>) -> usize {
    let at = models.len();
    models.push(Value::Null);
    let fields: Vec<Value> = schema["properties"]
        .as_object()
        .map(|members| {
            members
                .iter()
                .map(|(name, member)| {
                    json!({
                        "name": name,
                        "desc": member.get("description"),
                        // A Rust field is keyed on the wire by the one name serde writes, so
                        // there is no second spelling for an alias to carry.
                        "alias": Value::Null,
                        "type": node(member, models),
                    })
                })
                .collect()
        })
        .unwrap_or_default();
    models[at] = json!({ "doc": schema.get("description"), "fields": fields });
    at
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Every JSON-schema scalar name maps to the kind BAML prints, and an unnamed one is `any`.
    /// The `number` and `boolean` arms were both deletable without a test noticing.
    #[test]
    fn every_scalar_type_name_maps_to_its_kind() {
        let mut models = Vec::new();
        for (declared, kind) in [
            ("string", "str"),
            ("integer", "int"),
            ("number", "float"),
            ("boolean", "bool"),
        ] {
            assert_eq!(
                node(&json!({ "type": declared }), &mut models),
                json!({ "kind": kind }),
                "for {declared}"
            );
        }
        assert_eq!(
            node(&json!({}), &mut models),
            json!({ "kind": "named", "name": "any" }),
            "a schema naming no type describes anything"
        );
    }

    /// A `type` written as an array. One name is not a choice — it is that type, read through
    /// `declared_type`'s Array arm — while two names are, and the null arm is lifted into
    /// `optional` rather than becoming an arm of its own. Three mutants lived in this gap: the
    /// Array arm, its null filter, and `union_arms`' arity test.
    #[test]
    fn a_type_array_is_a_type_when_it_names_one_and_a_choice_when_it_names_two() {
        let mut models = Vec::new();
        assert_eq!(
            node(&json!({ "type": ["string"] }), &mut models),
            json!({ "kind": "str" }),
            "one name is not a union"
        );
        assert_eq!(
            node(&json!({ "type": ["string", "null"] }), &mut models),
            json!({ "kind": "union", "of": [{ "kind": "str" }], "optional": true }),
            "the null arm becomes optionality"
        );
        assert_eq!(
            node(&json!({ "type": ["string", "integer"] }), &mut models),
            json!({
                "kind": "union",
                "of": [{ "kind": "str" }, { "kind": "int" }],
                "optional": false,
            })
        );
    }
    use schemars::JsonSchema;

    fn tree<T: JsonSchema>() -> Value {
        json_field_reflection::<T>()
    }

    #[derive(JsonSchema)]
    #[allow(dead_code)]
    /// A gift idea.
    struct Gift {
        title: String,
        why: Option<String>,
        rank: u8,
    }

    #[test]
    fn a_struct_becomes_a_model_whose_members_keep_their_order() {
        let tree = tree::<Gift>();
        assert_eq!(tree["type"], json!({ "kind": "model", "model": 0 }));

        let model = &tree["models"][0];
        assert_eq!(model["doc"], "A gift idea.", "the doc comment travels");
        let names: Vec<&str> = model["fields"]
            .as_array()
            .expect("members")
            .iter()
            .map(|field| field["name"].as_str().expect("a name"))
            .collect();
        assert_eq!(
            names,
            ["title", "why", "rank"],
            "declaration order, not alphabetical"
        );
    }

    #[test]
    fn the_scalars_map_onto_their_own_kinds() {
        let model = &tree::<Gift>()["models"][0];
        assert_eq!(model["fields"][0]["type"], json!({ "kind": "str" }));
        assert_eq!(model["fields"][2]["type"], json!({ "kind": "int" }));
    }

    /// `Option<T>` is a choice of two of which one is null, and upstream renders that as the one
    /// type plus `or null` rather than as a union naming null.
    #[test]
    fn an_option_is_one_arm_marked_optional() {
        let model = &tree::<Gift>()["models"][0];
        assert_eq!(
            model["fields"][1]["type"],
            json!({ "kind": "union", "of": [{ "kind": "str" }], "optional": true })
        );
    }

    #[test]
    fn a_vec_becomes_a_list_of_whatever_it_holds() {
        #[derive(JsonSchema)]
        #[allow(dead_code)]
        struct Wrap {
            gifts: Vec<Gift>,
        }
        let tree = tree::<Wrap>();
        let outer = &tree["models"][0]["fields"][0]["type"];
        assert_eq!(outer["kind"], "list");
        assert_eq!(
            outer["of"],
            json!({ "kind": "model", "model": 1 }),
            "the element is a model of its own"
        );
        assert_eq!(tree["models"][1]["doc"], "A gift idea.");
    }

    /// A map and a record are both `object` in a schema; what tells them apart is that a map
    /// declares no members of its own.
    #[test]
    fn a_map_becomes_a_dict_rather_than_a_model() {
        #[derive(JsonSchema)]
        #[allow(dead_code)]
        struct Wrap {
            tags: std::collections::HashMap<String, i32>,
        }
        let field = &tree::<Wrap>()["models"][0]["fields"][0]["type"];
        assert_eq!(
            *field,
            json!({ "kind": "dict", "key": { "kind": "str" }, "value": { "kind": "int" } })
        );
    }

    #[test]
    fn a_unit_enum_becomes_the_set_of_its_members() {
        #[derive(JsonSchema)]
        #[allow(dead_code)]
        enum Mood {
            Calm,
            Wry,
        }
        #[derive(JsonSchema)]
        #[allow(dead_code)]
        struct Wrap {
            mood: Mood,
        }
        let field = &tree::<Wrap>()["models"][0]["fields"][0]["type"];
        assert_eq!(field["kind"], "literal");
        assert_eq!(field["members"], json!(["Calm", "Wry"]));
    }

    /// A field whose type says nothing about itself is named rather than given a shape it does
    /// not have.
    #[test]
    fn an_untyped_value_is_named_rather_than_shaped() {
        #[derive(JsonSchema)]
        #[allow(dead_code)]
        struct Wrap {
            anything: serde_json::Value,
        }
        let field = &tree::<Wrap>()["models"][0]["fields"][0]["type"];
        assert_eq!(*field, json!({ "kind": "named", "name": "any" }));
    }

    #[test]
    fn a_members_description_travels_with_it() {
        #[derive(JsonSchema)]
        #[allow(dead_code)]
        struct Wrap {
            /// who it is for
            recipient: String,
        }
        let field = &tree::<Wrap>()["models"][0]["fields"][0];
        assert_eq!(field["desc"], "who it is for");
    }
}
